docs/proposals/parallel-compaction.md

```diff
@@ -0,0 +1,68 @@
+weight: 1
+slug: parallel-compaction
+---
+
+- Author: [Roy Chiang](https://github.com/roystchiang)
+- Date: May 2021
+- Status: Proposed
+---
+
+## Introduction
+As a part of pushing Cortex’s scaling capability at AWS, we have done performance testing with Cortex and found the compactor to be one of the main limiting factors for higher active timeseries limit per tenant. The documentation [Compactor](https://cortexmetrics.io/docs/blocks-storage/compactor/#how-compaction-works) describes the responsibilities of a compactor, and this proposal focuses on the limitations of the current compactor architecture. In the current architecture, compactor has simple sharding, meaning that a single tenant is sharded to a single compactor. In addition, a compactor handles compaction groups of a single tenant iteratively, meaning that blocks belonging non-overlapping times are not compacted in parallel.
+
+### Problem and Requirements
+Currently, a compactor is able to compact up to 20M timeseries within 2 hours for a level-2 compaction, including the time to download blocks, compact, and upload the newly compacted block. We would like to increase the timeseries limit per tenant, and compaction is one of the limiting factors. In addition, we would like to achieve the following:
+
+* Compact multiple non-overlapping time intervals concurrently, so we can achieve higher throughput for the compaction of a single tenant
+* We should be able to scale up, down compactor as needed, depending on how many compactions are pending
+* Insight into the compaction progress of a tenant, such as the number of compactions required in order to catch up to the newest blocks
+
+## Design
+
+We accept the fact that a single compaction can potentially take more than 2 hours to compact, and we achieve higher compaction throughput through horizontally scaling the compactor. To compact more blocks in parallel for a single tenant, we distribute the compaction groups to compactors, instead of introducing more parallelism within a compactor.
+
+### Parallelize Work
+
+This proposal builds heavily on top of the [GrafanaLabs approach of introducing parallelism via time intervals](https://github.com/cortexproject/cortex/pull/2616). The difference being that a single tenant is now sharded across multiple compactors instead of just a single compactor. The initial approach will be to work on distinct time intervals, but the compactor planner can be later extended to introduce parallelism within a time interval as well. The
+
+The following is an example of parallelize work at each level:
+
+![Parallel Compaction Grouping](/images/proposals/parallel-compaction-grouping.png)
+
+Compactors are shuffle-sharded, meaning that 1 tenant can belong to multiple compactors, and these subset of compactors determine which blocks should be compacted together. Compactors determine amongst themselves the responsibility of the compaction blocks, by using a hash on time interval and the block ids.
```

Comment:
> by using a hash on time interval and the block ids

Why not just an hash of the tenant ID + time interval? I think block IDs shouldn't be part of the hash used for sharding purposes. We want to ensure that, given a specific time interval, only 1 compactor compacts blocks for that tenant and time interval. When 2 different compactors run the planning at a different time, the blocks they discover for the same tenant and time interval may be slightly different (think about a block uploaded by ingester with some delay) and they may end up both compacting it (because the generated hash is different). I think we have to prevent this situation.

Reply:
I originally thought about using the blocks hash because we control `consistency_delay`. However, now I noticed the default value is 0, and we shouldn't rely on this behavior to prevent multiple compactions across the same time interval.

I agree with your approach, and let's use a hash of tenant ID + time interval. The approach is simpler, and prevents the situation you described.